src/libraries/System.Net.Security/src/System/Net/Security/Pal.OSX/SafeDeleteSslContext.cs

```diff
@@ -7,35 +7,111 @@
+using System.Threading;
+using System.Threading.Tasks;
 using Microsoft.Win32.SafeHandles;
 
+using PAL_NwStatusUpdates = Interop.AppleCrypto.PAL_NwStatusUpdates;
+
+
 namespace System.Net
 {
     internal sealed class SafeDeleteSslContext : SafeDeleteContext
     {
+        public static readonly unsafe bool CanUseNwFramework = Interop.AppleCrypto.NwInit(&FramerStatusUpdate, &ReadFromConnection, &WriteToConnection) == 0;
+        public readonly bool UseNwFramework;
         // mapped from OSX error codes
         private const int OSStatus_writErr = -20;
         private const int OSStatus_readErr = -19;
+
+        private const int OSStatus_eofErr = - 39;
         private const int OSStatus_noErr = 0;
         private const int OSStatus_errSSLWouldBlock = -9803;
+
+        private const int OSStatus_errSecUserCanceled = -128;
         private const int InitialBufferSize = 2048;
         private readonly SafeSslHandle _sslContext;
         private ArrayBuffer _inputBuffer = new ArrayBuffer(InitialBufferSize);
         private ArrayBuffer _outputBuffer = new ArrayBuffer(InitialBufferSize);
 
+        public GCHandle gcHandle;
+
+        private  ManualResetEventSlim? _writeWaiter;
+        private int _writeStatus;
+        public ManualResetEventSlim? _readWaiter;
+        private int _readStatus;
+        public IntPtr _framer;
+        public bool handshakeStarted;
+
+        public TaskCompletionSource<SecurityStatusPalErrorCode>? Tcs;
+
         public SafeSslHandle SslContext => _sslContext;
         public SslApplicationProtocol SelectedApplicationProtocol;
         public bool IsServer;
 
-        public SafeDeleteSslContext(SslAuthenticationOptions sslAuthenticationOptions)
+        private bool _handshakeDone;
+        private bool _disposed;
+
+        public unsafe SafeDeleteSslContext(SslAuthenticationOptions sslAuthenticationOptions)
             : base(IntPtr.Zero)
         {
             try
             {
                 int osStatus;
 
-                _sslContext = CreateSslContext(sslAuthenticationOptions);
+                switch (sslAuthenticationOptions.EncryptionPolicy)
+                {
+                    case EncryptionPolicy.RequireEncryption:
+#pragma warning disable SYSLIB0040 // NoEncryption and AllowNoEncryption are obsolete
+
+                    case EncryptionPolicy.AllowNoEncryption:
+                        // SecureTransport doesn't allow TLS_NULL_NULL_WITH_NULL, but
+                        // since AllowNoEncryption intersect OS-supported isn't nothing,
+                        // let it pass.
+                        break;
+#pragma warning restore SYSLIB0040
+                    default:
+                        throw new PlatformNotSupportedException(SR.Format(SR.net_encryptionpolicy_notsupported, sslAuthenticationOptions.EncryptionPolicy));
+                }
+
+                // TBD make this opt-in
+                // NW freamewoprk still does not support all features and server side
```

Comment:
It supports only SslStream scenarios where you connect from a client to a server, not the other way around, ie. accepting TLS connections on the server side does not work (or rather works only through the deprecated CoreTLS API as previous .NET versions).

That generally boils down to the fact that the Framer API is abused to steal the TLS frames. It’s possible to make a fake connection to a bogus endpoint and then intercept all the data that would be sent to the connection. It’s not easily possible to fake that you accepted a connection from outside and intercept the data on that.

Reply:
Thanks. So "still does not support all features and server side" should be "still does not support all features nor accepting TLS connections server-side"?

Reply:
yes. I have some ideas for server but I'm running out of time at the moment. The goal at the moment is to support client connecting to TLS 1.3 they cannot control. It would be off by default for now but users who need it could subscribe to it in way similar to managed ntlm.